cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -389,9 +399,22 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
   @AuditPolicy(AuditDetail.REQUEST_BODY)
   public void upgradeApplication(HttpRequest request, HttpResponder responder,
                                  @PathParam("namespace-id") String namespaceId,
-                                 @PathParam("app-id") String appName) throws Exception {
-    ApplicationId appId = validateApplicationId(namespaceId, appName);
-    applicationLifecycleService.upgradeApplication(appId, createProgramTerminator());
+                                 @PathParam("app-id") String appName,
+                                 @Nullable @QueryParam("artifactScope") Set<String> artifactScopes,
```

Comment:
I think it will never be null if the type is a Collection. Please validate and remove the `@Nullable` if that's the case.